optuna/_gp/gp.py

```diff
@@ -103,6 +103,10 @@ def __init__(
         self._is_categorical = is_categorical
         self._X_train = X_train
         self._y_train = y_train
+        self._squared_X_diff = (X_train[..., None, :] - X_train[..., None, :, :]).square()
+        self._squared_X_diff[..., self._is_categorical] = (
+            self._squared_X_diff[..., self._is_categorical] > 0.0
+        ).type(torch.float64)
```

Comment:
How about limiting this only when `if self._is_categorical.any():`, to maintain consistency with the line 164:
https://github.com/nabenabe0928/optuna/blob/enhance/cache-pair-wise-distance/optuna/_gp/gp.py#L164
```suggestion
        if self._is_categorical.any():
            self._squared_X_diff[..., self._is_categorical] = (
                self._squared_X_diff[..., self._is_categorical] > 0.0
            ).type(torch.float64)
```